### PR TITLE
Update uploads via FTP

### LINKED ISSUE #10: Upload media files via rclone FTP
- [x] add a new setting called `UPLOADS_FTP_SERVER`
- [x] use wp hooks to prefixed uploads path on CRUD
- [x] decide wether to use ftp, based on `UPLOADS_FTP_SERVER`
- [x] remove GCS integration
- [ ] tests via rclone local

### PR DESCRIPTION
**Pull request by vtemian**
Uploads can now be served via FTP. 

Because PHP has native support for PHP stream wrappers, we add a filter to append `ftp://<host>:<port>` to each upload path.
We do that in order to support multiple uploads sources via rclone ftp server.

Fixes #10 

TODOs:
-  [x] test delete photo
-  [x] test delete pdf

#### Tests
* `git checkout uploads-via-ftp`
* in `.env` set `GCS_BUCKET`, `GCS_PROJECT_ID` and `GCS_SERVICE_ACCOUNT_JSON_KEY_PATH` to your local service account json key
* `docker-compose up`
* upload a photo (keep in mind that the docker-compose is using an old version of rclone)
* upload another photo
* you **should** see the second photo in the bucket, with the thumbnails generated
* delete the photo
* in the bucket, the photo and associated thumbnails **should** be deleted

### CODE REVIEW

File: src/Presslabs/FTPStorage.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+     *       https://core.trac.wordpress.org/ticket/38907#ticket
+     *
+     * Because path_join() doesn't recognize ftp:// as an absolute path, we need
+     * to remove ftp://<host>:<port>/ since for thumbnail will multiple such prefixes
+     * eg: ftp://<host>:<port>/wp-content/uploads/ftp://<host>:<port>/wp-content/uploads/...
+     */
+    public function filterDeleteFile($file) : string
+    {
+        $basedir = $this->getUploadsDir();
+        while (substr_count($file, $basedir) > 1) {
```
Comment (calind) [inline src/Presslabs/FTPStorage.php]: You should check that count > 1 and first occurence is at the begining of the string. Something like:
```php
if ($this->hasPrefix($file)) {
    while ($this->hasPrefix($file)) {
        $file = $this->stripPrefix($file);
    }
    $file = $this->getPrefix() . $file;
}

```

File: src/Presslabs/FTPStorage.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        $this->prefix = $prefix;
+    }
+
+    /**
+     *  Helper method that creates a new directory, using the current prefix.
+     *  Used in tests.
+     */
+    public function createDir(string $dir)
+    {
+        $recursive = substr_count($dir, '/') > 0;
```
Comment (calind) [inline src/Presslabs/FTPStorage.php]: Wouldn't  `ftp://` allways have more then 0 slashes?

Comment (vtemian) [reply src/Presslabs/FTPStorage.php]: Sure, but this one won't receive the full path, only the relative path.
And I need to know if it contains nested directories or just one, in order to pass recursive True/False.

File: src/Presslabs/FTPStorage.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        if ($this->startsWith($filePath, $baseDir)) {
+            while ($this->startsWith($filePath, $baseDir)) {
+                $filePath = $this->removePrefix($filePath, $baseDir);
+            }
+            $filePath = $baseDir . $filePath;
+        }
+
+        if ($this->startsWith($filePath, 'ftp://')) {
+            @unlink($filePath);
+            // unlink() does not clear the cache if you are performing filePath_exists()
```
Comment (calind) [inline src/Presslabs/FTPStorage.php]:
```suggestion
            // unlink() does not clear the cache if you are performing file_exists() on a remote file.
```

File: src/Presslabs/FTPStorage.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            // http://php.net/manual/en/function.clearstatcache.php#105888
+            clearstatcache();
```
Comment (calind) [inline src/Presslabs/FTPStorage.php]: We should return here to save a round trip.

**Pull request merged**